File: ros2_ws/src/my_cpp_pkg/src/number_publisher.cpp
#include <chrono>

#include "example_interfaces/msg/int64.hpp"
#include "rclcpp/rclcpp.hpp"

using namespace std::chrono_literals;

// Type alias for the Int64 message
using Integer64 = example_interfaces::msg::Int64;

class NumberPublisherNode : public rclcpp::Node {
public:
  NumberPublisherNode() : Node("number_publisher") {
    // ROS 2 parameters
    this->declare_parameter("count", 0);
    this->declare_parameter("timer_period", 1.0);

    count_ = this->get_parameter("count").as_int();
    timer_period_ = this->get_parameter("timer_period").as_double();

    param_callback_handle_ = this->add_post_set_parameters_callback(
      std::bind(&NumberPublisherNode::parametersCallback, this, std::placeholders::_1));

    publisher_ = this->create_publisher<Integer64>("number", 10);
    timer_ = this->create_wall_timer(
      std::chrono::duration<double>(timer_period_),
      std::bind(&NumberPublisherNode::publish_number, this));
    RCLCPP_WARN(this->get_logger(), "Number Publisher Node has started publishing numbers...");
  }

private:
  rclcpp::Publisher<Integer64>::SharedPtr publisher_;

  int64_t count_;
  double timer_period_;
  rclcpp::TimerBase::SharedPtr timer_;
  rclcpp::node_interfaces::NodeParametersInterface::OnSetParametersCallbackHandle::SharedPtr
    param_callback_handle_;
  int number_{};

  // Callback of timer to publish number
  void publish_number() {
    auto msg = Integer64();
    msg.data = count_;
    count_++;
    publisher_->publish(msg);
  }

  void parametersCallback(const std::vector<rclcpp::Parameter>& parameters) {
    for (const auto& param : parameters) {
      if (param.get_name() == "number") {
        number_ = param.as_int();
      }
    }
  }
};

int main(int argc, char** argv)
{
  rclcpp::init(argc, argv);
  auto node = std::make_shared<NumberPublisherNode>();
  rclcpp::spin(node);
  rclcpp::shutdown();
  return 0;
}

File: ros2_ws/src/my_python_pkg/my_python_pkg/number_publisher.py
#!/usr/bin/env python3
import rclpy
from rclpy.node import Node
from rclpy.parameter import Parameter
from example_interfaces.msg import Int64
 
 
class NumberPublisherNode(Node):
    def __init__(self):
        super().__init__("number_publisher") 

        # ROS2 Parameters below !
        self.declare_parameter("count", 0) # int
        self.declare_parameter("timer_period", 1.0) # float
        self.count_ = self.get_parameter("count").value
        self.timer_period_ = self.get_parameter("timer_period").value
        # Below line is for parameter callbacks and utilizing 'ros2 param set --- '
        self.add_post_set_parameters_callback(self.parameters_callback)
        
        self.publisher_ = self.create_publisher(Int64, "number", 10)
        self.timer_ = self.create_timer(self.timer_period_, self.publish_number_callback)   
        self.get_logger().warning("Python Number Publisher has started publishing...")

    def parameters_callback(self, params):
        for param in params:
            if param.name == "count":
                self.count_ = param.value
                self.get_logger().info(f"Count parameter updated to: {self.count_}")
            elif param.name == "timer_period":
                self.timer_period_ = param.value
                self.timer_.cancel()  # Cancel the existing timer
                self.timer_ = self.create_timer(self.timer_period_, self.publish_number_callback)  # Create a new timer with the updated period
                self.get_logger().info(f"Timer period parameter updated to: {self.timer_period_}")
        
    
    def publish_number_callback(self):
        msg = Int64()
        msg.data = self.count_
        self.publisher_.publish(msg)
        self.count_ += 1

 
def main(args=None):
    rclpy.init(args=args)
    node = NumberPublisherNode() 
    rclpy.spin(node)
    rclpy.shutdown()
 
 
if __name__ == "__main__":
    main()